presto-main/src/main/java/com/facebook/presto/cost/CapDistinctValuesCountToTypeDomainRangeLength.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.cost;
+
+import com.facebook.presto.spi.type.BigintType;
+import com.facebook.presto.spi.type.BooleanType;
+import com.facebook.presto.spi.type.IntegerType;
+import com.facebook.presto.spi.type.SmallintType;
+import com.facebook.presto.spi.type.TinyintType;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.Map;
+
+public class CapDistinctValuesCountToTypeDomainRangeLength
+        implements ComposableStatsCalculator.Normalizer
+{
+    @Override
+    public PlanNodeStatsEstimate normalize(PlanNode node, PlanNodeStatsEstimate estimate, Map<Symbol, Type> types)
+    {
+        PlanNodeStatsEstimate.Builder normalized = PlanNodeStatsEstimate.buildFrom(estimate);
+        for (Symbol symbol : estimate.getSymbolsWithKnownStatistics()) {
+            double domainLength = calculateDomainLength(symbol, estimate, types);
+            normalized.addSymbolStatistics(symbol, estimate.getSymbolStatistics(symbol)
+                    .mapDistinctValuesCount(distinctValuesCount -> Math.min(distinctValuesCount, domainLength)));
+        }
+
+        return normalized.build();
+    }
+
+    private double calculateDomainLength(Symbol symbol, PlanNodeStatsEstimate estimate, Map<Symbol, Type> types)
+    {
+        if (!isIntegralType(types.get(symbol))) {
+            return Double.NaN;
+        }
+
+        SymbolStatsEstimate symbolStatistics = estimate.getSymbolStatistics(symbol);
+        return symbolStatistics.getHighValue() - symbolStatistics.getLowValue() + 1;
```

Comment:
I think round down to whole value would be justifiable.
Also I think you can just `ceil(symbolStatistics.getHighValue() - symbolStatistics.getLowValue())` (without adding 1.